File: Java/src/main/java/com/gildedrose/handler/BackstagePassHandler.java
package com.gildedrose.handler;

import com.gildedrose.Item;
import com.gildedrose.utility.ItemUtilities;

public class BackstagePassHandler implements ItemHandler {
    private static final int DOUBLE_QUALITY_INCREASE_MAX_EXCLCUSIVE = 11;
    private static final int TRIPLE_QUALITY_INCREASE_MAX_EXCLUSIVE = 6;

    @Override
    public boolean handles(String itemName) {
        return "Backstage passes to a TAFKAL80ETC concert".equals(itemName);
    }

    @Override
    public void updateItem(Item item) {
        if (ItemUtilities.canIncreaseQuality(item)) {
            ItemUtilities.increaseQuality(item);

            if (ItemUtilities.isSellInLessThan(item, DOUBLE_QUALITY_INCREASE_MAX_EXCLCUSIVE) && ItemUtilities.canIncreaseQuality(item)) {
                ItemUtilities.increaseQuality(item);
            }

            if (ItemUtilities.isSellInLessThan(item, TRIPLE_QUALITY_INCREASE_MAX_EXCLUSIVE) && ItemUtilities.canIncreaseQuality(item)) {
                ItemUtilities.increaseQuality(item);
            }
        }

        ItemUtilities.decreaseSellIn(item);

        if (ItemUtilities.isExpired(item)) {
            ItemUtilities.resetQuality(item);
        }
    }
}

File: TypeScript/app/handler/BackstagePassHandler.ts
import { Item } from '../gilded-rose';
import { ItemHandler } from './ItemHandler';
import { ItemUtils } from '../utility/ItemUtils';

export class BackstagePassHandler implements ItemHandler {
  handles(item: Item): boolean {
    return item.name === 'Backstage passes to a TAFKAL80ETC concert';
  }

  updateItemQuality(item: Item): void {
    if (item.quality < 50) {
      ItemUtils.increaseQuality(item);
      if (item.sellIn < 11) {
        ItemUtils.increaseQuality(item);
      }
      if (item.sellIn < 6) {
        ItemUtils.increaseQuality(item);
      }
    }
    ItemUtils.decreaseSellIn(item);
    if (item.sellIn < 0) {
      item.quality = 0;
    } else {
      item.quality = Math.min(50, item.quality);
    }
  }
}
